### PR TITLE
[Merged by Bors] - feat: port Order.Interval

### PR DESCRIPTION
**Pull request by eric-wieser**
---

~~(Edit by @joneugster)~~
~~The issue about the wrongly generated `ext`-Lemmas for `NonemptyInterval` has been brought up on [zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/!4.233360.20new.20ext.20lemma.20generation/near/351104097). A suggestion is to extend `ext` by an option `@[ext (flat := false)]` in order to be able to toggle whether `ext` should flatten structures or not:~~

- [x] leanprover/std4#122

~~(2nd edit: the proposed change hasn't seen much progress. This PR can be merged without it and I would create a new PR fixing the porting note if the changes to `ext` would ever hit `Std`)~~

[Open in Gitpod](https://gitpod.io/from-referrer/)

Replaces #1272

### CODE REVIEW

**Comment by joneugster** [comment]
Now most of the remaining errors have something to do with the interplay of `NonemptyInterval.toProd` and `NonemptyInterval.toDualProd`. It seems that lean3 could somehow change one into the other, and here lean4 fails to do so. I dont really understand much about that

File: Mathlib/Order/Interval.lean [hunk 1]
```diff
@@ -731,22 +739,43 @@ noncomputable instance completeLattice [@DecidableRel α (· ≤ ·)] :
...
+                have h₂ := (WithBot.coe_le_coe.1 <| ha _ hc).2
+                have h := (s.fst_le_snd.trans h₂)
+                clear h₂
+                have h₃ : b.toProd.fst ≤ s.toProd.fst
+                · repeat rw [NonemptyInterval.toDualProd_apply] at h₁
+                  dsimp at h₁
+                  -- Porting note: (TODO): Why does this work the way it does.
+                  -- This next proof step looks somehow very sketchy to me!
+                  -- I beleive the API around duals might hav a problem.
+                  rw [OrderDual.le_def] at h₁
```
Comment (joneugster) [inline Mathlib/Order/Interval.lean:775]: This seems somewhat sketchy. I think something about the `OrderDual` API does not work as intended.

This step essentailly turns `toDual s.toProd.fst ≤ toDual b.toProd.fst` into `toDual b.toProd.fst ≤ toDual s.toProd.fst` and it is not obvious what hidden instances it changes in order to do so.

File: Mathlib/Order/Interval.lean [hunk 2]
```diff
@@ -673,12 +676,17 @@ section CompleteLattice
 
 variable [CompleteLattice α]
 
+-- Porting note: (TODO) The API around `OrderDual` seems very sketchy to me.
+-- I used this helper lemma in the proof below but, somehow it seems that's
+-- just an indication that the API does not work. In mathlib3, it would automatically do stuff
+-- (this lemma should probably be removed again, once things work)
+theorem OrderDual.le_def (x y : αᵒᵈ) [h : LE α] : (OrderDual.le α).le x y ↔ h.le y x := by rfl
```
Comment (eric-wieser) [inline Mathlib/Order/Interval.lean:683]: As noted on Zulip, we definitely don't want this lemma

File: Mathlib/Order/Interval.lean (hunk 1 above)
Comment (joneugster) [reply Mathlib/Order/Interval.lean]: short discussion about this on [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/!4.233360.20OrderDual).

File: Mathlib/Order/Interval.lean (hunk 2 above)
Comment (joneugster) [reply Mathlib/Order/Interval.lean]: done

**Comment by joneugster** [comment]
todo: I just uncommented all `@[simp, norm_cast]` attributes where the linter was not happy, search for `simpNF`. Usually it was because LHS `a ⊆b` got simplified to `a ≤ b` and I don't know what to do about them. Should maybe be fixed before merging.

**Comment by ChrisHughes24** [comment]
bors merge

**Pull request closed without merging**